File: engine/runtime/save_manager.py
from __future__ import annotations


"""
Copyright (c) 2026 Diego Millan - Pick A Path
Licensed under the Pick-A-Path Public License v1.0.
See LICENSE.txt in the project root for full license terms.
Commercial use without prior written consent is strictly prohibited.
"""

# engine/runtime/save_manager.py
"""
SaveManager - routes save/load operations to the correct backend.

Backends
--------
LocalStorageBackend  : browser-side localStorage (JS companion in save_manager.js)
FileBackend          : local filesystem (author kit, author's pick export, etc.)
ApiBackend           : stub for future hosted platform

SaveManager itself is Python-side only (server / offline). The browser-side
counterpart is save_manager.js, which talks to the Flask server via /save and
/load endpoints.

Usage (server_launcher.py / server.py)
---------------------------------------
    from engine.runtime.save_manager import SaveManager, FileBackend

    sm = SaveManager(FileBackend(game_root="path/to/game"))
    bundle = session.export_save_data(save_type="player_slot", display_name="My Save")
    sm.save("slot_1", bundle)
    bundle = sm.load("slot_1")
    result = session.load_from_data(bundle)
"""


import json
import os
import hashlib
from datetime import datetime
from typing import Optional

# ---------------------------------------------------------------------------
# Exceptions
# ---------------------------------------------------------------------------

class SaveError(Exception):
    pass


class LoadError(Exception):
    pass


class IncompatibleSaveError(LoadError):
    """Raised when a save cannot be loaded due to incompatible game version."""
    def __init__(self, message, at_node=None, hint=None):
        super().__init__(message)
        self.at_node = at_node
        self.hint = hint


# ---------------------------------------------------------------------------
# Backend interface
# ---------------------------------------------------------------------------

class _Backend:
    """Base interface all backends must implement."""

    def save(self, key: str, bundle: dict) -> bool:
        raise NotImplementedError

    def load(self, key: str) -> Optional[dict]:
        raise NotImplementedError

    def delete(self, key: str) -> bool:
        raise NotImplementedError

    def list_saves(self) -> list[dict]:
        """Returns lightweight summaries (no full telemetry_path) for slot UI."""
        raise NotImplementedError


# ---------------------------------------------------------------------------
# FileBackend  (offline author kit + author_pick export)
# ---------------------------------------------------------------------------

class FileBackend(_Backend):
    """
    Stores save bundles as JSON files under <game_root>/saves/.
    Also handles author_pick export to <game_root>/author_pick.json.

    Key naming convention:
      player saves  : "slot_1" ... "slot_10"
      author pick   : "author_pick"   → written to game root, not saves/
      checkpoints   : "checkpoint_<id>"
      bug reports   : "bug_<timestamp>"
    """

    SAVES_DIR = "saves"
    AUTHOR_PICK_FILE = "author_pick.json"

    def __init__(self, game_root: str):
        self.game_root = game_root
        self._saves_dir = os.path.join(game_root, self.SAVES_DIR)
        os.makedirs(self._saves_dir, exist_ok=True)

    def _path_for(self, key: str) -> str:
        if key == "author_pick":
            return os.path.join(self.game_root, self.AUTHOR_PICK_FILE)
        safe_key = key.replace("/", "_").replace("\\", "_")
        return os.path.join(self._saves_dir, f"{safe_key}.json")

    def save(self, key: str, bundle: dict) -> bool:
        path = self._path_for(key)
        try:
            with open(path, "w", encoding="utf-8") as f:
                json.dump(bundle, f, indent=2, ensure_ascii=False)
            return True
        except OSError as e:
            raise SaveError(f"Could not write save file '{path}': {e}") from e

    def load(self, key: str) -> Optional[dict]:
        path = self._path_for(key)
        if not os.path.exists(path):
            return None
        try:
            with open(path, "r", encoding="utf-8") as f:
                return json.load(f)
        except (OSError, json.JSONDecodeError) as e:
            raise LoadError(f"Could not read save file '{path}': {e}") from e

    def delete(self, key: str) -> bool:
        path = self._path_for(key)
        if os.path.exists(path):
            os.remove(path)
            return True
        return False

    def list_saves(self) -> list[dict]:
        summaries = []
        if not os.path.isdir(self._saves_dir):
            return summaries
        for fname in sorted(os.listdir(self._saves_dir)):
            if not fname.endswith(".json"):
                continue
            path = os.path.join(self._saves_dir, fname)
            try:
                with open(path, "r", encoding="utf-8") as f:
                    bundle = json.load(f)
                summaries.append(_bundle_summary(fname[:-5], bundle))
            except Exception:
                summaries.append({"key": fname[:-5], "error": "unreadable"})
        return summaries


# ---------------------------------------------------------------------------
# SaveManager
# ---------------------------------------------------------------------------

class SaveManager:
    """
    Routes save/load to the correct backend and applies pre/post-processing
    (checksum verification, type routing, summary stripping).

    Typical offline setup
    ---------------------
        sm = SaveManager(FileBackend(game_root="./my_game"))

    Typical future online setup
    ---------------------------
        sm = SaveManager(
            primary=ApiBackend(base_url="https://api.mygame.com", auth_token=token),
            fallback=FileBackend(game_root="./my_game"),
        )
    """

    # Maps save_type → default slot key prefix
    TYPE_PREFIXES = {
        "player_slot": "slot",
        "author_pick": "author_pick",
        "checkpoint":  "checkpoint",
        "bug_report":  "bug",
    }

    def __init__(self, primary: _Backend, fallback: Optional[_Backend] = None):
        self.primary = primary
        self.fallback = fallback

    # --- Core operations ---

    def save(self, key: str, bundle: dict) -> bool:
        """
        Write bundle to primary backend.
        Falls back to secondary on failure if configured.
        """
        try:
            return self.primary.save(key, bundle)
        except SaveError as e:
            if self.fallback:
                print(f"[SaveManager] Primary failed ({e}), trying fallback.")
                return self.fallback.save(key, bundle)
            raise

    def load(self, key: str) -> Optional[dict]:
        """
        Read bundle from primary backend, fall back if missing.
        Returns None if not found in either backend.
        """
        bundle = self.primary.load(key)
        if bundle is None and self.fallback:
            bundle = self.fallback.load(key)
        return bundle

    def delete(self, key: str) -> bool:
        ok = self.primary.delete(key)
        if self.fallback:
            self.fallback.delete(key)
        return ok

    def list_saves(self) -> list[dict]:
        return self.primary.list_saves()

    # --- Slot helpers ---

    def save_slot(self, slot_number: int, bundle: dict) -> bool:
        """Save to a numbered player slot (1–10)."""
        if not 1 <= slot_number <= 10:
            raise SaveError("Slot number must be between 1 and 10.")
        return self.save(f"slot_{slot_number}", bundle)

    def load_slot(self, slot_number: int) -> Optional[dict]:
        """Load from a numbered player slot."""
        return self.load(f"slot_{slot_number}")

    def save_author_pick(self, bundle: dict) -> bool:
        """
        Saves the author's pick bundle.
        Enforces save_type and is_read_only regardless of what bundle contains.
        """
        bundle = dict(bundle)
        bundle["save_type"] = "author_pick"
        bundle["is_read_only"] = True
        return self.save("author_pick", bundle)

    def load_author_pick(self) -> Optional[dict]:
        return self.load("author_pick")

    def save_bug_report(self, bundle: dict) -> bool:
        ts = datetime.now().strftime("%Y%m%d_%H%M%S")
        key = f"bug_{ts}"
        bundle = dict(bundle)
        bundle["save_type"] = "bug_report"
        return self.save(key, bundle)

    # --- Summary for UI ---

    def slot_summaries(self) -> list[dict]:
        """
        Returns a list of 10 slot entries for the save/load UI.
        Empty slots are represented with {"slot": N, "empty": True}.
        """
        results = []
        saves = {s["key"]: s for s in self.list_saves() if s.get("key", "").startswith("slot_")}
        for n in range(1, 11):
            key = f"slot_{n}"
            if key in saves:
                entry = dict(saves[key])
                entry["slot"] = n
                results.append(entry)
            else:
                results.append({"slot": n, "empty": True})
        return results


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _bundle_summary(key: str, bundle: dict) -> dict:
    """
    Strips the large telemetry_path and choice_tape from a bundle for
    lightweight slot-list display.
    """
    return {
        "key": key,
        "save_type": bundle.get("save_type", "player_slot"),
        "display_name": bundle.get("display_name", key),
        "game_id": bundle.get("game_id"),
        "game_version": bundle.get("game_version"),
        "created_at": bundle.get("created_at"),
        "is_read_only": bundle.get("is_read_only", False),
        "vars_snapshot": bundle.get("vars_snapshot", {}),
        "choice_count": len(bundle.get("choice_tape", [])),
        "checksum": bundle.get("checksum"),
    }

File: engine/web/js/save_manager.js
// engine/web/js/save_manager.js
//
// Browser-side SaveManager
// ========================
// Handles localStorage persistence for offline play, and will proxy to the
// server API for cloud saves once the hosted platform is built.
//
// Storage layout (localStorage keys)
// ------------------------------------
//   pq_save_slot_1  …  pq_save_slot_10   - player slots
//   pq_author_pick                       - author's pick (read-only)
//   pq_save_index                        - lightweight slot summary array
//
// The full bundle (with telemetry_path) is stored in the slot keys.
// The index stores only summaries (no large arrays) for fast slot-list rendering.

const SaveManager = (() => {

  const PREFIX       = "pq_save_";
  const INDEX_KEY    = "pq_save_index";
  const AUTHOR_KEY   = "pq_author_pick";
  const MAX_SLOTS    = 10;

  // -------------------------------------------------------------------------
  // Low-level storage helpers
  // -------------------------------------------------------------------------

  function _write(key, bundle) {
    try {
      localStorage.setItem(key, JSON.stringify(bundle));
      return true;
    } catch (e) {
      // QuotaExceededError or similar
      console.error("[SaveManager] localStorage write failed:", e);
      return false;
    }
  }

  function _read(key) {
    try {
      const raw = localStorage.getItem(key);
      return raw ? JSON.parse(raw) : null;
    } catch (e) {
      console.error("[SaveManager] localStorage read failed:", e);
      return null;
    }
  }

  function _delete(key) {
    localStorage.removeItem(key);
  }

  // -------------------------------------------------------------------------
  // Bundle summary (mirrors Python's _bundle_summary)
  // -------------------------------------------------------------------------

  function _summary(key, bundle) {
    return {
      key,
      save_type:    bundle.save_type    || "player_slot",
      display_name: bundle.display_name || key,
      game_id:      bundle.game_id      || null,
      game_version: bundle.game_version || null,
      created_at:   bundle.created_at   || null,
      is_read_only: bundle.is_read_only || false,
      vars_snapshot: bundle.vars_snapshot || {},
      choice_count: (bundle.choice_tape || []).length,
      checksum:     bundle.checksum     || null,
    };
  }

  // -------------------------------------------------------------------------
  // Index management  (keeps the slot UI fast)
  // -------------------------------------------------------------------------

  function _loadIndex() {
    return _read(INDEX_KEY) || {};
  }

  function _saveIndex(index) {
    _write(INDEX_KEY, index);
  }

  function _updateIndex(key, bundle) {
    const index = _loadIndex();
    index[key] = _summary(key, bundle);
    _saveIndex(index);
  }

  function _removeFromIndex(key) {
    const index = _loadIndex();
    delete index[key];
    _saveIndex(index);
  }

  // -------------------------------------------------------------------------
  // Core API
  // -------------------------------------------------------------------------

  /**
   * Save a bundle to a player slot (1–10).
   * @param {number} slotNumber  1–10
   * @param {Object} bundle      Full bundle from /api/save_data or built locally
   * @returns {boolean}
   */
  function saveSlot(slotNumber, bundle) {
    if (slotNumber < 1 || slotNumber > MAX_SLOTS) {
      console.error("[SaveManager] Slot number must be 1–10.");
      return false;
    }
    const key = `${PREFIX}slot_${slotNumber}`;
    const ok = _write(key, bundle);
    if (ok) _updateIndex(key, bundle);
    return ok;
  }

  /**
   * Load a player slot bundle.
   * @param {number} slotNumber  1–10
   * @returns {Object|null}
   */
  function loadSlot(slotNumber) {
    return _read(`${PREFIX}slot_${slotNumber}`);
  }

  /**
   * Delete a player slot.
   * @param {number} slotNumber  1–10
   */
  function deleteSlot(slotNumber) {
    const key = `${PREFIX}slot_${slotNumber}`;
    _delete(key);
    _removeFromIndex(key);
  }

  /**
   * Save the author's pick bundle.
   * @param {Object} bundle
   */
  function saveAuthorPick(bundle) {
    const stamped = { ...bundle, save_type: "author_pick", is_read_only: true };
    return _write(AUTHOR_KEY, stamped);
  }

  /**
   * Load the author's pick bundle.
   * @returns {Object|null}
   */
  function loadAuthorPick() {
    return _read(AUTHOR_KEY);
  }

  /**
   * Returns an array of 10 slot entries for the save/load UI.
   * Empty slots have { slot: N, empty: true }.
   */
  function slotSummaries() {
    const index = _loadIndex();
    const result = [];
    for (let n = 1; n <= MAX_SLOTS; n++) {
      const key = `${PREFIX}slot_${n}`;
      if (index[key]) {
        result.push({ slot: n, ...index[key] });
      } else {
        result.push({ slot: n, empty: true });
      }
    }
    return result;
  }

  // -------------------------------------------------------------------------
  // Server sync  (request/response via existing Flask endpoints)
  // -------------------------------------------------------------------------

  /**
   * Ask the server to build a save bundle for the current session state,
   * then persist it locally and optionally to the server (cloud stub).
   *
   * @param {number} slotNumber
   * @param {Object} options   { save_type, display_name, author_note }
   * @returns {Promise<{ok: boolean, bundle?: Object, error?: string}>}
   */
  async function requestSaveFromServer(slotNumber, options = {}) {
    try {
      const res = await fetch("/save", {
        method: "POST",
        headers: {
            'Content-Type': 'application/json',
            'X-CSRFToken': getCsrfToken()
        },
        body: JSON.stringify({
          slot: slotNumber,
          save_type:    options.save_type    || "player_slot",
          display_name: options.display_name || null,
          author_note:  options.author_note  || null,
        }),
      });
      if (!res.ok) {
        const err = await res.text();
        return { ok: false, error: err };
      }
      const bundle = await res.json();
      saveSlot(slotNumber, bundle);
      return { ok: true, bundle };
    } catch (e) {
      return { ok: false, error: e.message };
    }
  }

  /**
   * Load a slot bundle from localStorage and send it to the server to
   * reconstruct the session, then return the resulting game frame.
   *
   * @param {number} slotNumber
   * @returns {Promise<{ok: boolean, frame?: Object, warning?: string, error?: string}>}
   */
  async function requestLoadFromServer(slotNumber) {
    const bundle = loadSlot(slotNumber);
    if (!bundle) {
      return { ok: false, error: `Slot ${slotNumber} is empty.` };
    }
    try {
      const res = await fetch("/load", {
        method: "POST",
        headers: {
            'Content-Type': 'application/json',
            'X-CSRFToken': getCsrfToken()
        },
        body: JSON.stringify(bundle),
      });
      if (!res.ok) {
        const err = await res.text();
        return { ok: false, error: err };
      }
      const data = await res.json();
      if (data.error) return { ok: false, error: data.error, at_node: data.at_node };
      return {
        ok: true,
        frame:   data.frame,
        warning: data.message || null,   // version mismatch warning
      };
    } catch (e) {
      return { ok: false, error: e.message };
    }
  }

  /**
   * Request the author's pick bundle from the server (reads author_pick.json
   * from the game root) and cache it in localStorage.
   */
  async function fetchAuthorPickFromServer() {
    try {
      const res = await fetch("/author_pick");
      if (!res.ok) return { ok: false, error: "No author's pick found." };
      const bundle = await res.json();
      saveAuthorPick(bundle);
      return { ok: true, bundle };
    } catch (e) {
      return { ok: false, error: e.message };
    }
  }

  /**
   * Export a bug report, ask the server to build a bundle with save_type
   * "bug_report", then offer it as a file download.
   *
   * @param {string} userComment   Player's description of the bug
   */
  async function exportBugReport(userComment = "") {
    try {
      const res = await fetch("/save", {
        method: "POST",
        headers: {
            'Content-Type': 'application/json',
            'X-CSRFToken': getCsrfToken()
        },
        body: JSON.stringify({
          save_type: "bug_report",
          author_note: userComment,
        }),
      });
      if (!res.ok) return { ok: false, error: await res.text() };
      const bundle = await res.json();
      _downloadJSON(bundle, `bug_report_${Date.now()}.json`);
      return { ok: true };
    } catch (e) {
      return { ok: false, error: e.message };
    }
  }

  // -------------------------------------------------------------------------
  // Utilities
  // -------------------------------------------------------------------------

  /** Trigger a JSON file download in the browser. */
  function _downloadJSON(obj, filename) {
    const blob = new Blob([JSON.stringify(obj, null, 2)], { type: "application/json" });
    const url  = URL.createObjectURL(blob);
    const a    = document.createElement("a");
    a.href     = url;
    a.download = filename;
    a.click();
    URL.revokeObjectURL(url);
  }

  /** Load a .json save file from disk (file input element). */
  function loadFromFile(file) {
    return new Promise((resolve, reject) => {
      const reader = new FileReader();
      reader.onload = e => {
        try { resolve(JSON.parse(e.target.result)); }
        catch (err) { reject(new Error("Invalid save file.")); }
      };
      reader.onerror = () => reject(new Error("Could not read file."));
      reader.readAsText(file);
    });
  }

  // -------------------------------------------------------------------------
  // Public API
  // -------------------------------------------------------------------------

  return {
    // Sync (localStorage)
    saveSlot,
    loadSlot,
    deleteSlot,
    saveAuthorPick,
    loadAuthorPick,
    slotSummaries,

    // Async (server-backed)
    requestSaveFromServer,
    requestLoadFromServer,
    fetchAuthorPickFromServer,
    exportBugReport,

    // Utilities
    loadFromFile,
  };
})();